File: Example/iZettleSDKSample/ViewController/CardPaymentViewController.m
//
//  ViewController.m
//  iZettle SDK Sample
//
//  Copyright © 2015 PayPal Inc. All rights reserved.
//

#import "CardPaymentViewController.h"
#import "ConsoleViewController.h"

@import iZettleSDK;

@interface CardPaymentViewController ()
@property (weak, nonatomic) IBOutlet UILabel *amountLabel;
@property (weak, nonatomic) IBOutlet UIStepper *stepper;
@property (weak, nonatomic) IBOutlet UIToolbar *lastPaymentInfoToolBar;
@property (weak, nonatomic) IBOutlet UISwitch *tippingSwitch;
@property (weak, nonatomic) IBOutlet UILabel *selectedTippingStyleLabel;
@property (assign) IZSDKTippingStyle selectedTippingStyle;
@end

@implementation CardPaymentViewController {
    iZettleSDKPaymentInfo *_lastPaymentInfo;
    NSString *_lastReference;
    NSDate *_timestamp;
    NSError *_lastError;
}

#pragma mark - View lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];
    self.amountLabel.text = @"10.0";
    self.stepper.value = self.amountLabel.text.doubleValue;
    self.selectedTippingStyle = IZSDKTippingStyleMarketDefault;
    [self _updateTippingStyleLabelFromStyle:self.selectedTippingStyle];
}

- (IBAction)tippingStyleValueChanged:(id)sender {
    
    if (self.tippingSwitch.isOn) {
        __weak __auto_type wself = self;
        
        UIAlertController *tippingStyleSelectionSheet = [UIAlertController alertControllerWithTitle:@"Tipping style" message:@"Select prefered tipping style" preferredStyle:UIAlertControllerStyleActionSheet];
        
        UIAlertAction *marketDefaultAction = [UIAlertAction actionWithTitle:@"Market default" style:UIAlertActionStyleDefault handler:^(id _) {
            wself.selectedTippingStyle = IZSDKTippingStyleMarketDefault;
            [wself _updateTippingStyleLabelFromStyle:IZSDKTippingStyleMarketDefault];
        }];
        
        UIAlertAction *amountAction = [UIAlertAction actionWithTitle:@"Amount" style:UIAlertActionStyleDefault handler:^(id _) {
            wself.selectedTippingStyle = IZSDKTippingStyleAmount;
            [wself _updateTippingStyleLabelFromStyle:IZSDKTippingStyleAmount];
        }];
        
        UIAlertAction *percentageAction = [UIAlertAction actionWithTitle:@"Percentage" style:UIAlertActionStyleDefault handler:^(id _) {
            wself.selectedTippingStyle = IZSDKTippingStylePercentage;
            [wself _updateTippingStyleLabelFromStyle:IZSDKTippingStylePercentage];
        }];
        
        UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(id _) {
            [tippingStyleSelectionSheet dismissViewControllerAnimated:YES completion:nil];
            
            if (wself.selectedTippingStyle == IZSDKTippingStyleNone) {
                [wself.tippingSwitch setOn:NO];
            }
        }];
        
        [tippingStyleSelectionSheet addAction:marketDefaultAction];
        [tippingStyleSelectionSheet addAction:amountAction];
        [tippingStyleSelectionSheet addAction:percentageAction];
        [tippingStyleSelectionSheet addAction:cancelAction];
        
        [self presentViewController:tippingStyleSelectionSheet animated:YES completion:nil];
    } else {
        self.selectedTippingStyle = IZSDKTippingStyleNone;
        
        [self _updateTippingStyleLabelFromStyle:self.selectedTippingStyle];
    }
}

#pragma mark - User interaction

- (IBAction)stepperValueChanged:(UIStepper *)sender {
    _amountLabel.text = [NSString stringWithFormat: @"%.1f", sender.value];
}

- (IBAction)charge:(id)sender {
    // Note! Replace this reference with your own if you want to be able to query iZettle about payment info and perform refunds.
    _lastReference = [self _uniqueReference];
    NSDecimalNumber *amount = [NSDecimalNumber decimalNumberWithString: _amountLabel.text];
    
    [[iZettleSDK shared] chargeAmount:amount
                         tippingStyle:self->_selectedTippingStyle
                            reference:_lastReference
            presentFromViewController:self
                           completion:^(iZettleSDKPaymentInfo *paymentInfo, NSError *error) {
        self->_lastPaymentInfo = paymentInfo;
        self->_lastError = error;
        self->_timestamp = [NSDate date];
        
        [self _updateLastPaymentToolbar];
        [self _showConsole];
    }];
}

- (IBAction)lastPaymentInfo:(id)sender {
    [[iZettleSDK shared] retrievePaymentInfoForReference:_lastReference presentFromViewController:self completion:^(iZettleSDKPaymentInfo *paymentInfo, NSError *error) {
        self->_lastPaymentInfo = paymentInfo;
        self->_lastError = error;
        [self _showConsole];
    }];
}

- (IBAction)refundLastPayment:(id)sender {
    NSString *paymentReference = _lastReference;
    _lastReference = [self _uniqueReference];
    
    [[iZettleSDK shared] refundAmount:nil ofPaymentWithReference:paymentReference refundReference:_lastReference presentFromViewController:self completion:^(iZettleSDKPaymentInfo * _Nullable paymentInfo, NSError * _Nullable error) {
        self->_lastPaymentInfo = paymentInfo;
        self->_lastError = error;
        [self _showConsole];
    }];
}

#pragma mark - Segue handling

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([[segue identifier] isEqualToString: @"showConsole"] && sender == self) {
        ConsoleViewController *console = segue.destinationViewController;
        if (_lastError) {
            [console addToLog:_lastError];
        } else {
            [console addToLog:_lastPaymentInfo];
        }
    }
}

#pragma mark - Helpers

- (void)_updateLastPaymentToolbar {
    _lastPaymentInfoToolBar.hidden = !(_lastPaymentInfo && _lastPaymentInfo.description);
    if (!_lastPaymentInfoToolBar.hidden) {
        UIBarButtonItem *lastPaymentButton = [_lastPaymentInfoToolBar.items firstObject];
        NSDateFormatter *dateFormatter = [NSDateFormatter new];
        dateFormatter.dateStyle = NSDateFormatterNoStyle;
        dateFormatter.timeStyle = NSDateFormatterMediumStyle;
        lastPaymentButton.title = [NSString stringWithFormat:@"Last payment, %@", [dateFormatter stringFromDate:_timestamp]];
    }
}

- (void)_showConsole {
    [self performSegueWithIdentifier:@"showConsole" sender:self];
}

- (NSString *)_uniqueReference {
    return [[NSUUID UUID] UUIDString];
}

- (void)_updateTippingStyleLabelFromStyle:(IZSDKTippingStyle)style {
    NSString *name = [self _tippingStyleNameFromStyle:style];
    
    [self.selectedTippingStyleLabel setText:name];
}

- (NSString *)_tippingStyleNameFromStyle:(IZSDKTippingStyle)style {
    switch (style) {
        case IZSDKTippingStyleNone: return @"None";
        case IZSDKTippingStyleMarketDefault: return @"Market default style";
        case IZSDKTippingStyleAmount: return @"Amount style";
        case IZSDKTippingStylePercentage: return @"Percentage style";
    }
}

@end

File: Example/iZettleSDKSample/ViewController/CardPaymentViewController.swift
//
//  CardPaymentViewController.swift
//  iZettle SDK Sample
//
//  Copyright © 2024 PayPal Inc. All rights reserved.
//

import UIKit
import iZettleSDK

class CardPaymentViewController: UIViewController {
    
    @IBOutlet weak var amountLabel: UILabel!
    @IBOutlet weak var stepper: UIStepper!
    @IBOutlet weak var tippingSettingsButton: UIButton!
    @IBOutlet weak var tippingSwitch: UISwitch!
    @IBOutlet weak var lastPaymentInfoToolBar: UIToolbar!
    
    private var lastPaymentInfo: iZettleSDKPaymentInfo?
    private var lastReference: String?
    private var timestamp: Date?
    private var lastError: Error?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tippingSettingsButton.setTitle("", for: .normal)
        amountLabel.text = "10.0"
        stepper.value = Double(amountLabel.text ?? "0") ?? 0.0
    }
    
    
    // MARK: - User Interaction
    
    @IBAction func tippingStyleValueChanged(_ sender: UISwitch) {
        
        if tippingSwitch.isOn {
            tippingSettingsButton.isEnabled = true
        } else {
            tippingSettingsButton.isEnabled = false
        }
    }
    
    
    @IBAction func tippingSettingsButtonTapped(_ sender: UIButton) {
        let tippingSettingsVC = UINavigationController(rootViewController: TippingSettingsViewController())
        present(tippingSettingsVC, animated: true)
    }
    
    @IBAction func stepperValueChanged(_ sender: UIStepper) {
        amountLabel.text = String(format: "%.1f", sender.value)
    }
    
    @IBAction func charge(_ sender: Any) {
        lastReference = makeUniqueRefrence()
        let amount = NSDecimalNumber(string: amountLabel.text)
        
        iZettleSDK.shared().charge(
            amount: amount,
            tippingConfiguration: getSelectedTippingConfiguration(),
            reference: lastReference,
            presentFrom: self
        ) { [weak self] paymentInfo, error in
            self?.lastPaymentInfo = paymentInfo
            self?.lastError = error
            self?.timestamp = Date()
            
            self?.updateLastPaymentToolbar()
            self?.showConsole()
        }
    }
    
    @IBAction func lastPaymentInfo(_ sender: Any) {
        if let lastReference {
            iZettleSDK.shared().retrievePaymentInfo(for: lastReference, presentFrom: self) { [weak self] paymentInfo, error in
                self?.lastPaymentInfo = paymentInfo
                self?.lastError = error
                self?.showConsole()
            }
        }
    }
    
    @IBAction func refundLastPayment(_ sender: Any) {
        let paymentReference = lastReference
        lastReference = makeUniqueRefrence()
        
        if let paymentReference {
            iZettleSDK.shared().refund(amount: nil, ofPayment: paymentReference, withRefundReference: lastReference, presentFrom: self) { [weak self] paymentInfo, error in
                self?.lastPaymentInfo = paymentInfo
                self?.lastError = error
                self?.showConsole()
            }
        }
        
    }
    
    // MARK: - Segue
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "showConsole" && sender as? UIViewController ==  self {
            if let console = segue.destination as? ConsoleViewController {
                if let lastError = lastError {
                    console.add(toLog: lastError)
                } else {
                    console.add(toLog: lastPaymentInfo)
                }
            }
        }
    }
    
    // MARK: - Helpers
    private func makeUniqueRefrence() -> String {
        return UUID().uuidString
    }
    
    func getSelectedTippingConfiguration() -> IZSDKTippingConfiguration {

        guard tippingSwitch.isOn else {
            return .disabled()
        }

        let zettleReaderTippingStyle = IZZettleReaderTippingStyle(rawValue: TippingSettingsStorage.shared.zettleReaderTippingStyle) ?? .none
        let paypalReaderTippingStyle = IZPayPalReaderTippingStyle(rawValue: TippingSettingsStorage.shared.paypalReaderTippingStyle) ?? .none

        if let storedPaypalReaderPredefinedTips = TippingSettingsStorage.shared.paypalReaderPredefinedTips {
            return IZSDKTippingConfiguration(
                zettleReaderTippingStyle: zettleReaderTippingStyle,
                paypalReaderTippingStyle: paypalReaderTippingStyle,
                paypalReaderPredefinedTippingValues: IZSDKPredefinedTippingValues(
                    option1: storedPaypalReaderPredefinedTips[0],
                    option2: storedPaypalReaderPredefinedTips[1],
                    option3: storedPaypalReaderPredefinedTips[2]
                )
            )
        } else {
            return IZSDKTippingConfiguration(
                zettleReaderTippingStyle: zettleReaderTippingStyle,
                paypalReaderTippingStyle: paypalReaderTippingStyle
            )
        }
    }
    
    private func showConsole() {
        performSegue(withIdentifier: "showConsole", sender: self)
    }
    
    private func updateLastPaymentToolbar() {
        lastPaymentInfoToolBar.isHidden = !((lastPaymentInfo != nil) && ((lastPaymentInfo?.description) != nil))
        
        if !(lastPaymentInfoToolBar.isHidden), let timestamp {
            let lastPaymentBarButton = lastPaymentInfoToolBar.items?.first
            let dateFormatter = DateFormatter()
            dateFormatter.dateStyle = .none
            dateFormatter.timeStyle = .medium
            lastPaymentBarButton?.title = "Last payment, \(dateFormatter.string(from: timestamp))"
        }
    }
    
}
